### PR TITLE
Implement extra DMX Clients

### PR DESCRIPTION
**Pull request by Breina**
Not sure if you're up for it, but it can't hurt to try.

- Separate animation logic from protocol logic
- Change date format to ISO 8601
- Implement KiNet (thanks @jjhuff)
- Implement e1.31 sACN (thanks @RichUncleCody)
- Thread is started automatically, no need to manually .start()

-----

I know the repo is named after ArtNet, but I'm using your lib in [my HA branch](https://github.com/Breina/ha-artnet-led) and would like it to have some more compatibility. I like your channel-width support and animation curves, prefer to not have to re-implement that separately.

Noticed your repo maintains a high standard; I did my best to match it. Please let me know how I did! :)

### CODE REVIEW

**Comment by Breina** [comment]
That should be all for now, would love to have this included into the 0.9.0 release.

File: pyartnet/dmx_client.py [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+
+        packet = bytearray()
+        # Root layer
+        packet.extend([0x00, 0x01, 0x00, 0x00])  # Preamble Size, Post-amble Size
+        packet.extend(map(ord, "ASC-E1.17"))  # Packet Identifier
+        packet.extend([0x00, 0x00, 0x00, 0x72, 0x57])  # padding x 3 , Flags, Length
+        packet.extend(struct.pack(">l", 4))  # Root Layer Vector
+        packet.extend(map(ord, "ThisIsMyCIDxxxxx"))  # CID, a unique identifier
+        # Framing layer
+        packet.extend([0x72, 0x57])  # Flags and length
```
Comment (spacemanspiff2007) [inline pyartnet/dmx_client.py:200]: Why is the length set to a static value? Shouldn't it be dynamic based on the highest channel?

File: pyartnet/dmx_client.py [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+    def update(self, universe_nr, universe):
+        """
+        Send the current state of DMX values to the gateway via UDP packet.
+        """
+
+        packet = self.__base_packet[:111]
+
+        packet.extend(struct.pack(">B", self.__sequence))
+        self.__sequence += 1
+        if self.__sequence == 200:
```
Comment (spacemanspiff2007) [inline pyartnet/dmx_client.py:218]: Why is it 200?

File: pyartnet/dmx_channel.py [hunk 3]
```diff
@@ -104,18 +104,19 @@ def add_fade(self, target_values: Iterable[Union[int, FadeBase]],
             self.__fades[i] = fade
 
         if log.isEnabledFor(logging.DEBUG):
-            log._log(logging.DEBUG, f'Fade with {self.__step_max} steps:', [])
+            log.log(logging.DEBUG, f'Fade with {self.__step_max} steps:', [])
```
Comment (spacemanspiff2007) [inline pyartnet/dmx_channel.py:107]: _log was intentional because it's checking the log level in the line above

File: pyartnet/animation_node.py [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+            return None
+        self.__task = asyncio.create_task(self.__worker())
+
+    async def stop(self):
+        if not self.__task:
+            return None
+
+        self.__task.cancel()
+
+        with contextlib.suppress(asyncio.CancelledError):
```
Comment (spacemanspiff2007) [inline pyartnet/animation_node.py:86]: Why do you use this?

**Comment by Breina** [comment]
Thanks for the feedback! I'll rework.

File: pyartnet/dmx_channel.py (hunk 3 above)
Comment (Breina) [reply pyartnet/dmx_channel.py]: Reverted

**Comment by spacemanspiff2007** [comment]
I'm all up for more features but time is short atm.

> I know the repo is named after ArtNet, but I'm using your lib in [my HA branch](https://github.com/Breina/ha-artnet-led) and would like it to have some more compatibility. I like your channel-width support and animation curves, prefer to not have to re-implement that separately.

I'd love to have more options but I'm currently very busy. So if you want these things working in the next couple of weeks I'd suggest you implement them for yourself and not wait for the PRs to be merged.
If you have time I'll gladly look at the things and we'll get them merged.

Also I think it makes sense to split things in multiple PRs, especially the new protocol support.

File: pyartnet/dmx_client.py (hunk 1 above)
Comment (Breina) [reply pyartnet/dmx_client.py:200]: That's only the length of the Framing layer, which is constant. The actual size of the universe data is hardcoded on line 224. (` packet.extend([0x72, 0x0d,`)

The synchronization mechanism of the E1.31 protocol seems to send the entire universe at once, so no highest channel can be selected as far as I've seen.

[ANSI_E1-31-2018.pdf](https://github.com/spacemanspiff2007/PyArtNet/files/9131501/ANSI_E1-31-2018.pdf)

File: pyartnet/dmx_client.py (hunk 2 above)
Comment (Breina) [reply pyartnet/dmx_client.py]: Increased to 255.

**Pull request closed without merging**